pkg/querier/queryrange/series_volume.go

```diff
@@ -0,0 +1,216 @@
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase/definitions"
+	"github.com/grafana/loki/pkg/util"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/user"
+)
+
+func SeriesVolumeDownstreamHandler(nextRT http.RoundTripper, codec queryrangebase.Codec) queryrangebase.Handler {
+	return queryrangebase.HandlerFunc(func(ctx context.Context, req queryrangebase.Request) (queryrangebase.Response, error) {
+		request, err := codec.EncodeRequest(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := user.InjectOrgIDIntoHTTPRequest(ctx, request); err != nil {
+			return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+		}
+
+		resp, err := nextRT.RoundTrip(request)
+		if err != nil {
+			return nil, err
+		}
+
+		return codec.DecodeResponse(ctx, resp, req)
+	})
+}
+
+func NewSeriesVolumeMiddleware() queryrangebase.Middleware {
+	return queryrangebase.MiddlewareFunc(func(next queryrangebase.Handler) queryrangebase.Handler {
+		return queryrangebase.HandlerFunc(func(ctx context.Context, req queryrangebase.Request) (queryrangebase.Response, error) {
+			volReq, ok := req.(*logproto.VolumeRequest)
+
+			if !ok {
+				return next.Do(ctx, req)
+			}
+
+			reqs := map[time.Time]queryrangebase.Request{}
+			startTS := volReq.From.Time()
+			endTS := volReq.Through.Time()
+			interval := time.Duration(volReq.Step * 1e6)
+
+			util.ForInterval(interval, startTS, endTS, true, func(start, end time.Time) {
+				// Range query buckets are aligned to the starting timestamp
+				// Instant queries are for "this instant", which aligns to the end of the requested range
+				bucket := start
+				if interval == 0 {
+					bucket = end
+				}
+
+				reqs[bucket] = &logproto.VolumeRequest{
+					From:     model.TimeFromUnix(start.Unix()),
+					Through:  model.TimeFromUnix(end.Unix()),
+					Matchers: volReq.Matchers,
+					Limit:    volReq.Limit,
+					Step:     volReq.Step,
+				}
+			})
+
+			type f func(context.Context) (time.Time, definitions.Response, error)
+			var jobs []f
+
+			for bucket, req := range reqs {
+				b, r := bucket, req
+				jobs = append(jobs, f(func(ctx context.Context) (time.Time, definitions.Response, error) {
+					resp, err := next.Do(ctx, r)
+					if err != nil {
+						return b, nil, err
+					}
+
+					return b, resp, nil
+				}))
+			}
+
+			collector := newConcurrentVolumeResponseCollector(len(reqs))
+			err := concurrency.ForEachJob(
+				ctx,
+				len(jobs),
+				len(jobs),
+				func(ctx context.Context, i int) error {
+					bucket, resp, err := jobs[i](ctx)
+					collector.Add(bucket, resp.(*VolumeResponse))
+					return err
+				})
+
+			if err != nil {
+				return nil, err
+			}
+
+			promResp := toPrometheusResponse(collector.responses)
+			return promResp, nil
+		})
+	})
+}
+
+type concurrentVolumeResponseCollector struct {
```

Comment:
If we're going to use a type for collecting responses like this, we need to be sure and protect `responses` on read, as well. I know we never call `collector.responses` before we're done but, as is, this a foot-gun for someone in the future.

Another approach you could take is a channel of 
```go
struct {
  bucket ...
  response ...
}
```

This would leave the `toPrometheusResponse` code generally unchanged and might be simpler?